flink-java/src/main/java/org/apache/flink/api/java/RemoteEnvironment.java

```diff
@@ -111,58 +93,59 @@ public RemoteEnvironment(String host, int port, Configuration clientConfig, Stri
-		}
-
-		this.host = host;
-		this.port = port;
-		this.clientConfiguration = clientConfig == null ? new Configuration() : clientConfig;
-		if (jarFiles != null) {
-			this.jarFiles = new ArrayList<>(jarFiles.length);
-			for (String jarFile : jarFiles) {
-				try {
-					this.jarFiles.add(new File(jarFile).getAbsoluteFile().toURI().toURL());
-				} catch (MalformedURLException e) {
-					throw new IllegalArgumentException("JAR file path invalid", e);
-				}
-			}
-		}
-		else {
-			this.jarFiles = Collections.emptyList();
-		}
-
-		if (globalClasspaths == null) {
-			this.globalClasspaths = Collections.emptyList();
-		} else {
-			this.globalClasspaths = Arrays.asList(globalClasspaths);
-		}
+	public RemoteEnvironment(String host, int port, Configuration clientConfig, String[] jarFiles, URL[] globalClasspaths) {
+		super(validateAndGetEffectiveConfiguration(clientConfig, host, port, jarFiles, globalClasspaths));
 	}
 
-	// ------------------------------------------------------------------------
+	private static Configuration validateAndGetEffectiveConfiguration(
+			final Configuration configuration,
+			final String host,
+			final int port,
+			final String[] jarFiles,
+			final URL[] globalClasspaths) {
+		RemoteEnvironmentConfigUtils.validate(host, port);
+		return getEffectiveConfiguration(
+				getClientConfiguration(configuration),
+				host,
+				port,
+				jarFiles,
+				getClasspathURLs(globalClasspaths));
+	}
 
-	@Override
-	public JobExecutionResult execute(String jobName) throws Exception {
-		final Plan p = createProgramPlan(jobName);
+	private static Configuration getClientConfiguration(final Configuration configuration) {
+		return configuration == null ? new Configuration() : configuration;
+	}
 
-		final PlanExecutor executor = PlanExecutor.createRemoteExecutor(host, port, clientConfiguration);
-		lastJobExecutionResult = executor.executePlan(p, jarFiles, globalClasspaths);
-		return lastJobExecutionResult;
+	private static List<URL> getClasspathURLs(final URL[] classpaths) {
+		return classpaths == null ? Collections.emptyList() : Arrays.asList(classpaths);
+	}
+
+	private static Configuration getEffectiveConfiguration(
+			final Configuration baseConfiguration,
+			final String host,
+			final int port,
+			final String[] jars,
+			final List<URL> classpaths) {
+
+		final Configuration effectiveConfiguration = new Configuration(baseConfiguration);
+
+		RemoteEnvironmentConfigUtils.setJobManagerAddressToConfig(host, port, effectiveConfiguration);
+		RemoteEnvironmentConfigUtils.setJarURLsToConfig(jars, effectiveConfiguration);
+		ConfigUtils.encodeCollectionToConfig(effectiveConfiguration, PipelineOptions.CLASSPATHS, classpaths, URL::toString);
+
+		// these should be set in the end to overwrite any values from the client config provided in the constructor.
+		effectiveConfiguration.setString(DeploymentOptions.TARGET, "remote-cluster-executor");
+		effectiveConfiguration.setBoolean(DeploymentOptions.ATTACHED, true);
```

Comment:
Why is `ATTACHED` the default?

Reply:
These were the previous semantics of the `RemoteExecutor`.